crates/ra_analysis/tests/tests.rs

```diff
@@ -144,3 +146,103 @@ fn test_resolve_crate_root() {
         vec![CrateId(1)],
     );
 }
+
+#[test]
+fn test_fn_signature() {
+    let snap = analysis(&[("/lib.rs",
+r#"fn foo(x: u32, y: u32) -> u32 {x + y}
+fn bar() { foo(3, ); }"#)]);
+
+    let (_handle, token) = JobHandle::new();
+
+
+    let offset = 53.into(); // (<|>3, )
+    let (desc, param) = snap.resolve_callable(FileId(1), offset, &token).unwrap();
+
+    assert_eq!(desc.name, "foo");
+    assert_eq!(desc.params, Some(vec!("x".into(),"y".into())));
+    assert_eq!(desc.ret_type, Some("-> u32".into()));
+    assert_eq!(param, Some(0));
+
+    let offset = 56.into(); // (3, <|>)
+    let (desc, param) = snap.resolve_callable(FileId(1), offset, &token).unwrap();
+
+    assert_eq!(desc.name, "foo");
+    assert_eq!(desc.params, Some(vec!("x".into(),"y".into())));
+    assert_eq!(desc.ret_type, Some("-> u32".into()));
+    assert_eq!(param, Some(1));
+}
+
+#[test]
+fn test_fn_signature_for_impl() {
+    let snap = analysis(&[("/lib.rs",
+r#"struct F; impl F { pub fn new() { F{}} }
+fn bar() {let _ : F = F::new();}"#)]);
+
+    let (_handle, token) = JobHandle::new();
+
+    let offset = 70.into(); // F::new(<|>)
+    let (desc, param) = snap.resolve_callable(FileId(1), offset, &token).unwrap();
+
+    assert_eq!(desc.name, "new");
+    assert_eq!(desc.params, None);
+    assert_eq!(desc.ret_type, None);
+    assert_eq!(param, None);
+}
+
+#[test]
+fn test_fn_signature_for_method_self() {
+    let snap = analysis(&[("/lib.rs",
+r#"struct F;
+impl F {
+    pub fn new() -> F{
+        F{}
+    }
+
+    pub fn do_it(&self) {}
+}
+
+fn bar() {
+    let f : F = F::new();
+    f.do_it();
+}"#)]);
+
+    let (_handle, token) = JobHandle::new();
+
+    let offset = 140.into(); // f.do_it(<|>)
+    let (desc, param) = snap.resolve_callable(FileId(1), offset, &token).unwrap();
+
+    assert_eq!(desc.name, "do_it");
+    assert_eq!(desc.params, Some(vec!["&self".to_string()]));
+    assert_eq!(desc.ret_type, None);
+    assert_eq!(param, None);
+}
+
+#[test]
+fn test_fn_signature_for_method_with_arg() {
+    let snap = analysis(&[("/lib.rs",
+r#"struct F;
+impl F {
+    pub fn new() -> F{
+        F{}
+    }
+
+    pub fn do_it(&self, x: i32) {}
+}
+
+fn bar() {
+    let f : F = F::new();
+    f.do_it();
+}"#)]);
+
+    let (_handle, token) = JobHandle::new();
+
+    let offset = 148.into(); // f.do_it(<|>)
+    let (desc, param) = snap.resolve_callable(FileId(1), offset, &token).unwrap();
+
+    assert_eq!(desc.name, "do_it");
+    assert_eq!(desc.params, Some(vec!["&self".to_string(), "x".to_string()]));
+    assert_eq!(desc.ret_type, None);
+    assert_eq!(param, Some(1));
+}
+}
```

Comment:
There are some copy-paste in this tests.

Perhaps we should have `fn get_signature()` which abstracts creation of an analysis with a single file and calling `resolve_callable`? 

Te call site should look like this:

```
fn test_fn_signature_for_method_self() {
    let (desc, param) = get_signature(
r#"struct F;
impl F {
    pub fn new() -> F{
        F{}
    }
    pub fn do_it(&self, x: i32) {}
}
fn bar() {
    let f : F = F::new();
    f.do_it(<|>);
}"#
)
}

    assert_eq!(desc.name, "do_it");
    assert_eq!(desc.params, Some(vec!["&self".to_string(), "x".to_string()]));
    assert_eq!(desc.ret_type, None);
    assert_eq!(param, Some(1));
```

Reply:
I think this is a good idea. I ran into some lifetime issues before but will try again this afternoon.